### DISCUSSION TITLE
Delete screenshots which trying to upload

### DISCUSSION THREAD

**Discussion by Parabak** [fastlane/fastlane#20858] [Ideas]
Hey guys,

From my point of view current `upload_screenshots` implementation in `deliver` action is ambiguous. 
Let's consider the case when I want to upload & overwrite screenshots for one particular locale e.g. `en-GB`.
p.s. _This option is vital In case you have ~1k screenshots you don't want to reupload them all as it took time and is failing from time to time due to AppStore issues_

I set `overwrite_screenshots: true` and run deliver:
```ruby
deliver(
      force: true,
      screenshots_path: "AppStore_upload",
      run_precheck_before_submit: false,
      overwrite_screenshots: true
    )
```

  `DeleteScreenshotSetJob` is enqueued for locales that are gonna be changed (in our case it's en-GB), which is nice. 
```ruby
  iterator = AppScreenshotIterator.new(localizations)
      iterator.each_app_screenshot_set do |localization, app_screenshot_set|
        # Only delete screenshots if trying to upload
        next unless screenshots_per_language.keys.include?(localization.locale)

        UI.verbose("Queued delete screenshot set job for #{localization.locale} #{app_screenshot_set.screenshot_display_type}")
        worker.enqueue(DeleteScreenshotSetJob.new(app_screenshot_set, localization))
      end

      worker.start
```
Once worker is finished all screenshots from `en-GB` are deleted 👍 

However, right after that, there is a check for all undeleted screenshots
```ruby
      # Verify all screenshots have been deleted
      # Sometimes API requests will fail but screenshots will still be deleted
      count = iterator.each_app_screenshot_set.map { |_, app_screenshot_set| app_screenshot_set }
                      .reduce(0) { |sum, app_screenshot_set| sum + app_screenshot_set.app_screenshots.size }
```
Voila, we have N screenshots undeleted — exit with error code. 
```shell
...
INFO [2022-11-15 15:04:37.40]: Deleted 'en-GB APP_IPAD_PRO_3GEN_129' -  (2.304244 secs)
WARN [2022-11-15 15:04:47.33]: Number of screenshots not deleted: 686
ERROR [2022-11-15 15:04:47.33]: Failed to delete all screenshots... Tries remaining: 4
...
```

I suggest to
* either verify that only enqueued screenshots were deleted 
* or replace bool `overwrite_screenshots` with some `enum: all, selected` where 
   * `all`  screenshots are deleted only for all Store locales versions. Deletion is verified
   * `selected` screenshots are deleted only for locales from `screenshots_path`. Deletion is verified only for those locales (not all)

The workaround I'm using now:

```
deliver(
      force: true,
      screenshots_path: "AppStore_upload", // pass only required locales
      run_precheck_before_submit: false,
      overwrite_screenshots: true
    )
```
All screenshots from required locales are deleted, but the action is failed. Repeast with `overwrite_screenshots` to just uploaded these locales 🤷 
```
deliver(
      force: true,
      screenshots_path: "AppStore_upload", // pass only required locales
      run_precheck_before_submit: false,
      overwrite_screenshots: false
    )
```

**Comment by jianwen-wang** [comment]
I encountered the same problem, and up to this point, it remains consistent with what Parabak mentioned.

**Comment by ralphcode** [comment]
Could this be solved by only verifying those being uploaded are now zero? Seems like an easy fix that won't change the overwrite behaviour. 

i.e. 
```
  # Verify all relevant screenshots have been deleted
  # Filter the screenshots to only include the ones being uploaded
  remaining_screenshots = iterator.each_app_screenshot_set
                                  .select { |localization, _| screenshots_per_language.keys.include?(localization.locale) }
                                  .map { |_, app_screenshot_set| app_screenshot_set }
                                  .reduce(0) { |sum, app_screenshot_set| sum + app_screenshot_set.app_screenshots.size }
```

**Comment by turushan** [comment]
Still an issue on 2.225.0
